### PR TITLE
Add OrderdProduct module

### LINKED ISSUE #59: Add OrderedProduct module
- Create OrderedProduct  module - fields and features.
- Write a test data.
- Write a pytest tests. 

Done on PR #58

### PR DESCRIPTION
**Pull request by MeitarRizner**
Add OrderdProduct module with all the features and tests. 

Fixes #59 

All the classes in models file are temporary except OrderedProduct (These classes didn't merge yet).

*Please let me know about missing features and missing tests*

### CODE REVIEW

**Comment by pazMenachem** [comment]
since im not exactly sure which basic tests the mentors/maintainers wants for module, the code looks good to me.
also assuming all the base class (user supplier client etc) will be deleted once those classes will be merged.

**Comment by Omeramsc** [comment]
Why does this PR also creates other modules? is it a WIP? or depends on another PR?

File: ordered_product/migrations/0001_initial.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
+# Generated by Django 4.1.3 on 2022-12-15 18:01
+
```
Comment (thetaxas2468) [inline ordered_product/migrations/0001_initial.py:2]: Delete comment

File: ordered_product/models.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    user_name = models.ForeignKey(Supplier, on_delete=models.CASCADE)
+    price = models.IntegerField(validators=[MinValueValidator(1)])
+    quantity = models.IntegerField(validators=[MinValueValidator(1)])
+
+    class Meta:
+        unique_together = ('suppler_product_id', 'qr_code')
+
+
+class OrderdProduct(models.Model):
+    ordered_product_id = models.IntegerField(primary_key=True, editable=False, unique=True)
```
Comment (thetaxas2468) [inline ordered_product/models.py:65]: Django creates auto field primary key,delete this one

File: ordered_product/models.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        self.save()
+        try:
+            OrderdProduct.full_clean(self)
+        except ValidationError as e:
+            self.delete()
+            raise e
+        return self
+
+    def delete_ordered_product(self):
+        OrderdProduct.objects.get(ordered_product_id=self.ordered_product_id).delete()
```
Comment (thetaxas2468) [inline ordered_product/models.py:101]: You have increase,decrease,order(with quantity = 1) functions that needs to needs to be here and missing,Look at the DFD

**Comment by thetaxas2468** [comment]
link this PR to an issue
check out :
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

File: ordered_product/models.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    qr_code = models.ForeignKey(Product, on_delete=models.CASCADE)
+    user_name = models.ForeignKey(Supplier, on_delete=models.CASCADE)
+    price = models.IntegerField(validators=[MinValueValidator(1)])
+    quantity = models.IntegerField(validators=[MinValueValidator(1)])
+
+    class Meta:
+        unique_together = ('suppler_product_id', 'qr_code')
+
+
+class OrderdProduct(models.Model):
```
Comment (yftacherzog) [inline ordered_product/models.py:64]:
```suggestion
class OrderedProduct(models.Model):
```

File: ordered_product/models.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    ordered_product_id = models.IntegerField(primary_key=True, editable=False, unique=True)
+    delivery_location_id = models.ForeignKey(DeliveryLocation, on_delete=models.CASCADE)
+    user_name = models.ForeignKey(Client, on_delete=models.CASCADE)
+    supplier_product_id = models.ForeignKey(SupplierProduct, on_delete=models.CASCADE)
+    quantity = models.IntegerField(default=1, validators=[MinValueValidator(1)])
+
+    @staticmethod
+    def filter_ordered_prodct_ID(opID):
```
Comment (yftacherzog) [inline ordered_product/models.py:72]: change variable and function name to snake_case with lowercase only. Also for the rest of the PR.

File: ordered_product/tests.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+@pytest.fixture
+def delivery_location0():
+    return DeliveryLocation(user_name=supplier0,
+                            location="Kiryat Shmona",
+                            date=datetime.date(2022, 12, 30))
+
+
+@pytest.fixture
+def ordered_product0():
```
Comment (yftacherzog) [inline ordered_product/tests.py:50]: You're creating fixtures for (supplier, client, etc.) but you're not using them. I was expecting to see them being used in this fixture instead of creating yet more instances. You can also save each of them from within those fixtures rather than cluttering this one.

File: ordered_product/tests.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+@pytest.fixture
+def saved_ordered_product0(ordered_product0):
+    ordered_product0.save_ordered_product()
+    return ordered_product0
+
+
+class TestOrderedProductModel():
+    @pytest.mark.django_db()
+    def test_save_and_delete_ordered_product(self, saved_ordered_product0):
+        temp = OrderdProduct(ordered_product_id=saved_ordered_product0.ordered_product_id,
```
Comment (yftacherzog) [inline ordered_product/tests.py:94]: You can just use the same OrderedProduct again and keep the test cleaner.

File: ordered_product/tests.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+    @pytest.mark.django_db()
+    def test_filter_by_ordered_prodct_ID(self, saved_ordered_product0):
+        assert saved_ordered_product0 not in OrderdProduct.filter_ordered_prodct_ID(1000)
+        filt = OrderdProduct.filter_ordered_prodct_ID(saved_ordered_product0.ordered_product_id)
+        assert [saved_ordered_product0] == list(filt)
+
+    @pytest.mark.django_db()
+    def test_filter_by_user_name(self, saved_ordered_product0):
+        assert saved_ordered_product0 not in OrderdProduct.filter_user_name("")
```
Comment (yftacherzog) [inline ordered_product/tests.py:113]: Why do we need to `list()` the result at the positive test but not on the negative test?

File: ordered_product/tests.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        filt = OrderdProduct.filter_user_name(saved_ordered_product0.user_name)
+        assert saved_ordered_product0 in list(filt)
```
Comment (yftacherzog) [inline ordered_product/tests.py:115]: Why using `in` here and `==` on the previous one? I think `==` is more appropriate (had a similar discussion with @thetaxas2468)

**Comment by thetaxas2468** [comment]
> link this PR to an issue check out : https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

Still not linked checkout the link, it should be something like this:
Commit information with an opened issue
<9 options> the issue id

Example:
Information
fixes #59

**Comment by MeitarRizner** [comment]
> Why does this PR also creates other modules? is it a WIP? or depends on another PR?

Its depends on another PR.

File: ordered_product/tests.py (hunk 6 above)
Comment (MeitarRizner) [reply ordered_product/tests.py:50]: I tried this (use fixture as instance) and it didn't work. I will try again

Comment (yftacherzog) [reply ordered_product/tests.py:50]: Let us know what's not working and we can try and figure this out together.

File: ordered_product/admin.py [hunk 10]
```diff
@@ -0,0 +1,9 @@
+from django.contrib import admin
+from ordered_product.models import OrderdProduct
+
+
+class OrderedProductAdmin(admin.ModelAdmin):
+    list_display = ('ordered_product_id', 'delivery_location_id', 'user_name', 'supplier_product_id', 'quantity', )
+
+
+admin.site.register(OrderdProduct, OrderedProductAdmin)
```
Comment (pazMenachem) [inline ordered_product/admin.py:9]:
```suggestion
admin.site.register(OrderedProduct, OrderedProductAdmin)
```

File: ordered_product/admin.py [hunk 11]
```diff
@@ -0,0 +1,9 @@
+from django.contrib import admin
+from ordered_product.models import OrderdProduct
```
Comment (pazMenachem) [inline ordered_product/admin.py:2]:
```suggestion
from ordered_product.models import OrderedProduct
```

File: ordered_product/migrations/0002_test_data.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
+from django.db import migrations, transaction
+from ordered_product.models import Supplier, Client, Product, SupplierProduct, DeliveryLocation, OrderdProduct
```
Comment (pazMenachem) [inline ordered_product/migrations/0002_test_data.py:2]:
```suggestion
from ordered_product.models import Supplier, Client, Product, SupplierProduct, DeliveryLocation, OrderedProduct
```

**Review by pazMenachem — COMMENTED** [review]
take note that User abstract class wont have user type field any more, therefor change the relevant sections in your code that include Supplier/Client.

File: ordered_product/models.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
+from django.db import models
+from django.core.validators import MinLengthValidator
+from django.core.validators import MinValueValidator
+from django.core.exceptions import ValidationError
+from product.models import Product
+
+
+class User(models.Model):
```
Comment (Omeramsc) [inline ordered_product/models.py:8]: Are all the classes in this file are temporary? because we don't need to redefine them.
if so, please mention it in the description.

File: ordered_product/models.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    def filter_delivery_location_id(dlID):
+        return OrderdProduct.objects.filter(delivery_location_id=dlID) if dlID else []
+
+    @staticmethod
+    def filter_supplier_product_id(spID):
+        return OrderdProduct.objects.filter(supplier_product_id=spID) if spID else []
+
+    @staticmethod
+    def get_all_ordered_product():
+        return OrderdProduct.objects.all()
```
Comment (Omeramsc) [inline ordered_product/models.py:89]: no need for this wrapper function, it doesn't add anything.

File: ordered_product/tests.py [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    def test_filter_by_supplier_product_id(self, saved_ordered_product0):
+        assert saved_ordered_product0 not in OrderdProduct.filter_supplier_product_id(-1)
+        filt = OrderdProduct.filter_supplier_product_id(saved_ordered_product0.supplier_product_id)
+        assert saved_ordered_product0 in list(filt)
+
+    @pytest.mark.django_db()
+    def test_filter_delivery_location_id(self, saved_ordered_product0):
+        assert saved_ordered_product0 not in OrderdProduct.filter_delivery_location_id(-1)
+        filt = OrderdProduct.filter_delivery_location_id(saved_ordered_product0.delivery_location_id)
+        assert saved_ordered_product0 in list(filt)
```
Comment (thetaxas2468) [inline ordered_product/tests.py:127]: test your quantity validator

File: ordered_product/models.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+
+class OrderedProduct(models.Model):
+    delivery_location_id = models.ForeignKey(DeliveryLocation, on_delete=models.CASCADE)
+    user_name = models.ForeignKey(Client, on_delete=models.CASCADE)
+    supplier_product_id = models.ForeignKey(SupplierProduct, on_delete=models.CASCADE)
+    quantity = models.IntegerField(default=1, validators=[MinValueValidator(1)])
+
+    @staticmethod
+    def filter_id(opID):
```
Comment (yftacherzog) [inline ordered_product/models.py:66]: snake_case

File: ordered_product/models.py [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+        return OrderedProduct.objects.filter(id=opID) if opID else None
```
Comment (yftacherzog) [inline ordered_product/models.py:67]: snake_case

File: ordered_product/models.py [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+
+    @staticmethod
+    def filter_user_name(userName):
```
Comment (yftacherzog) [inline ordered_product/models.py:70]: snake_case

File: ordered_product/models.py [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+        return OrderedProduct.objects.filter(user_name=userName) if userName else []
+
+    @staticmethod
+    def filter_delivery_location_id(dlID):
```
Comment (yftacherzog) [inline ordered_product/models.py:74]: snake_case

File: ordered_product/models.py [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+        return OrderedProduct.objects.filter(delivery_location_id=dlID) if dlID else []
+
+    @staticmethod
+    def filter_supplier_product_id(spID):
```
Comment (yftacherzog) [inline ordered_product/models.py:78]: snake_case

File: ordered_product/models.py (hunk 16 above)
Comment (yftacherzog) [inline ordered_product/models.py:66]: Can't we use `get` here? Do we need filter functionality for this field?

File: ordered_product/tests.py [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+        delivery_location_id=delivery_location0,
+        user_name=client0,
+        supplier_product_id=supplier_product0,
+        quantity=3)
+
+
+class TestOrderedProductModel():
+    @pytest.mark.django_db()
+    def test_save_and_delete_ordered_product(self, delivery_location0, supplier_product0, ordered_product0):
+        delivery_location0.save()
```
Comment (yftacherzog) [inline ordered_product/tests.py:60]: It would make sense to assert `not in` before adding.

File: ordered_product/tests.py [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+        supplier_product0.save()
+        ordered_product0.save_ordered_product()
+        assert ordered_product0 in OrderedProduct.objects.all()
+        ordered_product0.delete_ordered_product()
+        assert ordered_product0 not in OrderedProduct.objects.all()
+
+    @pytest.mark.django_db()
+    def test_filter_by_id(self, delivery_location0, supplier_product0, ordered_product0):
+        delivery_location0.save()
+        supplier_product0.save()
```
Comment (yftacherzog) [inline ordered_product/tests.py:70]: All your tests start with those 2 lines. I would make sense to put that inside a fixture.

File: ordered_product/tests.py [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+        ordered_product0.save_ordered_product()
+        assert ordered_product0 not in list(OrderedProduct.filter_id(1000))
+        filt = OrderedProduct.filter_id(ordered_product0.id)
+        assert ordered_product0 in list(filt)
```
Comment (yftacherzog) [inline ordered_product/tests.py:74]: For all those tests, `==` will make better sense comparing to `in`. This way, you will assert that you have exactly what you were filtering for and not other stuff.

File: ordered_product/tests.py [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+        assert ordered_product0 not in list(OrderedProduct.filter_delivery_location_id(-1))
+        filt = OrderedProduct.filter_delivery_location_id(ordered_product0.delivery_location_id)
+        assert ordered_product0 in list(filt)
+
+    @pytest.mark.django_db()
+    def test_validators(self, delivery_location0, supplier_product0, ordered_product0):
+        try:
+            delivery_location0.save()
+            supplier_product0.save()
+            ordered_product0.save_ordered_product()
```
Comment (yftacherzog) [inline ordered_product/tests.py:108]: Why are we saving this one?

File: ordered_product/tests.py [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+            tmp_ordered_product = ordered_product0
+            tmp_ordered_product.quantity = 0
+            tmp_ordered_product.save_ordered_product()
+        except ValidationError:
+            assert tmp_ordered_product not in OrderedProduct.objects.all()
```
Comment (yftacherzog) [inline ordered_product/tests.py:113]: Is this where we plan to fail? What's actually failing here?

File: ordered_product/tests.py [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+    @pytest.mark.django_db()
+    def test_decrease_quantity(self, delivery_location0, supplier_product0, ordered_product0):
+        delivery_location0.save()
+        supplier_product0.save()
+        ordered_product0.save_ordered_product()
+        ordered_product0.decrease_quantity(1)
+        assert ordered_product0.quantity == 2
+        assert ordered_product0 in OrderedProduct.objects.all()
+        ordered_product0.decrease_quantity(8)
+        assert ordered_product0.quantity < 0
```
Comment (yftacherzog) [inline ordered_product/tests.py:124]: A lot of magic numbers here, which makes the test more fragile and less transparent. I'd suggest either having all values set here (e.g. assign 3, decrease by 1, assert 2, and so on), or make the values proportional to existing value (e.g. store the current value to a local var, decrease by 1, assert equals temp-1, decrease by temp+8, assert deleted, etc.).

File: ordered_product/tests.py [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+        assert ordered_product0 not in OrderedProduct.objects.all()
+
+    @pytest.mark.django_db()
+    def test_increase_quantity(self, delivery_location0, supplier_product0, ordered_product0):
+        delivery_location0.save()
+        supplier_product0.save()
+        ordered_product0.save_ordered_product()
+        ordered_product0.increase_quantity(6)
+        assert ordered_product0.quantity == 9
```
Comment (yftacherzog) [inline ordered_product/tests.py:133]: same comment about magic numbers.

File: ordered_product/tests.py [hunk 28]
```diff
@@ -0,0 +1,147 @@
...
+@pytest.fixture
+def delivery_location0(supplier0):
+    return DeliveryLocation(user_name=supplier0,
+                            location="Kiryat Shmona",
+                            date=datetime.date(2022, 12, 30))
+
+
+@pytest.fixture
+def ordered_product0(delivery_location0, client0, supplier_product0):
+    return OrderedProduct(
```
Comment (thetaxas2468) [inline ordered_product/tests.py:50]: In here your tests pass because in your 0002 test file you have the same product,supplier,supplierproduct,deliverylocation (if you change the fields value i think all your tests will fail)

so before returning the ordered product save each one of the parameters deliver_location0....... by calling the functions that we have created in our models and after that return the OrderedProduct

File: ordered_product/tests.py [hunk 29]
```diff
@@ -0,0 +1,147 @@
...
+        assert ordered_product0 in OrderedProduct.objects.all()
+        ordered_product0.delete_ordered_product()
+        assert ordered_product0 not in OrderedProduct.objects.all()
+
+    @pytest.mark.django_db()
+    def test_filter_by_id(self, delivery_location0, supplier_product0, ordered_product0):
+        delivery_location0.save()
+        supplier_product0.save()
+        ordered_product0.save_ordered_product()
+        assert ordered_product0 not in list(OrderedProduct.filter_id(1000))
```
Comment (thetaxas2468) [inline ordered_product/tests.py:72]: What are you testing here?

File: ordered_product/tests.py [hunk 30]
```diff
@@ -0,0 +1,147 @@
...
+        filt = OrderedProduct.filter_id(ordered_product0.id)
+        assert ordered_product0 in list(filt)
+
+    @pytest.mark.django_db()
+    def test_filter_by_user_name(self, delivery_location0, supplier_product0, ordered_product0):
+        delivery_location0.save()
+        supplier_product0.save()
+        ordered_product0.save_ordered_product()
+        assert ordered_product0 not in list(OrderedProduct.filter_user_name(""))
```
Comment (thetaxas2468) [inline ordered_product/tests.py:81]: What are you testing here?

File: ordered_product/tests.py [hunk 31]
```diff
@@ -0,0 +1,147 @@
...
+        filt = OrderedProduct.filter_user_name(ordered_product0.user_name)
+        assert ordered_product0 in list(filt)
+
+    @pytest.mark.django_db()
+    def test_filter_by_supplier_product_id(self, delivery_location0, supplier_product0, ordered_product0):
+        delivery_location0.save()
+        supplier_product0.save()
+        ordered_product0.save_ordered_product()
+        assert ordered_product0 not in list(OrderedProduct.filter_supplier_product_id(-1))
```
Comment (thetaxas2468) [inline ordered_product/tests.py:90]: this should be in its own test function i guess

File: ordered_product/tests.py [hunk 32]
```diff
@@ -0,0 +1,147 @@
...
+        filt = OrderedProduct.filter_supplier_product_id(ordered_product0.supplier_product_id)
+        assert ordered_product0 in list(filt)
+
+    @pytest.mark.django_db()
+    def test_filter_delivery_location_id(self, delivery_location0, supplier_product0, ordered_product0):
+        delivery_location0.save()
+        supplier_product0.save()
+        ordered_product0.save_ordered_product()
+        assert ordered_product0 not in list(OrderedProduct.filter_delivery_location_id(-1))
```
Comment (thetaxas2468) [inline ordered_product/tests.py:99]: this should be in its own test function i guess

**Comment by thetaxas2468** [comment]
Please rebase your commit on top of upstream/main and fix your migrations because supplier is already merged and deliverylocation is about to be merged also and fix your paths in order look at my review on other PRs,Rawan and Lior

File: ordered_product/models.py (hunk 16 above)
Comment (MeitarRizner) [reply ordered_product/models.py:66]: You are right. I'll fix it.

File: ordered_product/models.py [hunk 33]
```diff
@@ -0,0 +1,105 @@
...
+        self.quantity -= qn
+        self.save()
+        if (self.quantity <= 0):
+            self.delete_ordered_product()
+
+    def increase_quantity(self, qn):
+        self.quantity += qn
+
+    @staticmethod
+    def order(delivery, client, supplier_product, qn):
```
Comment (pazMenachem) [inline ordered_product/models.py:103]: maybe check whether qn is 0, since we agreed product with 0 quantity is not valid.
would suggest raising an error if qn is 0 or create other indication the instance hasnt been saved.

**Review by pazMenachem — COMMENTED** [review]
there are methods that do the same as built in methods, no point warping those methods as  i see it.

File: ordered_product/tests.py (hunk 25 above)
Comment (MeitarRizner) [reply ordered_product/tests.py:113]: I'll fix it.

File: ordered_product/models.py (hunk 33 above)
Comment (thetaxas2468) [reply ordered_product/models.py:103]: if qn <= 0 and not only 0

File: ordered_product/tests.py (hunk 28 above)
Comment (MeitarRizner) [reply ordered_product/tests.py:50]: I'm using the supplier0 that Paz wrote and the tests still pass.. So I don't need to do these changes.

File: ordered_product/tests.py (hunk 29 above)
Comment (MeitarRizner) [reply ordered_product/tests.py:72]: I deleted this filter

File: ordered_product/tests.py (hunk 30 above)
Comment (MeitarRizner) [reply ordered_product/tests.py:81]: I created a new fixture 'client1' to filter by this fixture instead of ("")

File: ordered_product/tests.py (hunk 31 above)
Comment (MeitarRizner) [reply ordered_product/tests.py:90]: Why? I'm don't understand

File: ordered_product/tests.py (hunk 32 above)
Comment (MeitarRizner) [reply ordered_product/tests.py:99]: Why? I'm don't understand

File: ordered_product/models.py (hunk 33 above)
Comment (MeitarRizner) [reply ordered_product/models.py:103]: Okay

**Comment by MeitarRizner** [comment]
> there are methods that do the same as built in methods, no point warping those methods as i see it.

Which methods?

File: ordered_product/tests.py [hunk 34]
```diff
@@ -0,0 +1,90 @@
...
+    def test_validators(self, delivery_location0, client0, supplier_product0, save_objects):
+        try:
+            temp_ordered_product = OrderedProduct(
+                delivery_location_id=delivery_location0,
+                user_name=client0,
+                supplier_product_id=supplier_product0,
+                quantity=-2)
+            temp_ordered_product.save_ordered_product()
+        except ValidationError:
+            assert temp_ordered_product not in OrderedProduct.objects.all()
```
Comment (yftacherzog) [inline ordered_product/tests.py:55]: You should not have try/except blocks inside your tests. You should instead use the `with pytest.raises(ValidationError)` pattern. It's already implemented in several other PRs, so have a look.

File: ordered_product/tests.py [hunk 35]
```diff
@@ -0,0 +1,90 @@
...
+
+    @pytest.mark.django_db()
+    def test_order(self, client0, client1, ordered_product0, delivery_location0, supplier_product0, save_objects):
+        ordered_product0.user_name = client0
+        client1.save()
+        ordered_product0.save_ordered_product()
+        try:
+            ordered_product = OrderedProduct.order(delivery_location0, client1, supplier_product0, -1)
+        except ValidationError:
+            ordered_product = OrderedProduct.order(delivery_location0, client1, supplier_product0, 3)
```
Comment (yftacherzog) [inline ordered_product/tests.py:84]: Use `with pytest.raises(...)` instead of `try/except`.

File: ordered_product/tests.py [hunk 36]
```diff
@@ -0,0 +1,88 @@
...
+        assert ordered_product0.quantity == 9
+        assert ordered_product0 in OrderedProduct.objects.all()
+
+    @pytest.mark.django_db()
+    def test_order(self, client0, client1, ordered_product0, delivery_location0, supplier_product0, save_objects):
+        ordered_product0.user_name = client0
+        client1.save()
+        ordered_product0.save_ordered_product()
+        with pytest.raises(ValidationError):
+            ordered_product = OrderedProduct.order(delivery_location0, client1, supplier_product0, -1)
```
Comment (yftacherzog) [inline ordered_product/tests.py:81]: The negative scenario should be in a test case of its own.

File: ordered_product/tests.py [hunk 37]
```diff
@@ -0,0 +1,88 @@
...
+    @pytest.mark.django_db()
+    def test_validators(self, delivery_location0, client0, supplier_product0, save_objects):
+        with pytest.raises(ValidationError):
+            temp_ordered_product = OrderedProduct(
+                delivery_location_id=delivery_location0,
+                user_name=client0,
+                supplier_product_id=supplier_product0,
+                quantity=-2)
+            temp_ordered_product.save_ordered_product()
+        assert temp_ordered_product not in OrderedProduct.objects.all()
```
Comment (yftacherzog) [inline ordered_product/tests.py:54]: make the context manager include only the line that you expect to generate the exception. This way you increase the chances for actually spotting the error you were expecting.
```suggestion
        temp_ordered_product = OrderedProduct(
                delivery_location_id=delivery_location0,
                user_name=client0,
                supplier_product_id=supplier_product0,
                quantity=-2)
        with pytest.raises(ValidationError):
            temp_ordered_product.save_ordered_product()
        assert temp_ordered_product not in OrderedProduct.objects.all()
```

File: ordered_product/tests.py [hunk 38]
```diff
@@ -0,0 +1,88 @@
...
+
+    @pytest.mark.django_db()
+    def test_decrease_quantity(self, ordered_product0, save_objects):
+        ordered_product0.quantity = 3
+        ordered_product0.save_ordered_product()
+        ordered_product0.decrease_quantity(1)
+        assert ordered_product0.quantity == 2
+        assert ordered_product0 in OrderedProduct.objects.all()
+        ordered_product0.decrease_quantity(8)
+        assert ordered_product0.quantity < 0
```
Comment (Omeramsc) [inline ordered_product/tests.py:64]: what do we test here?

Comment (MeitarRizner) [reply ordered_product/tests.py:64]: > what do we test here?

When clients decrease quantity by more than they ordered so the order needs to deleted (if the quantity less or equals to zero --> delete order)

File: ordered_product/tests.py [hunk 39]
```diff
@@ -0,0 +1,93 @@
...
+        assert ordered_product0 in OrderedProduct.objects.all()
+
+    @pytest.mark.django_db()
+    def test_order(self, client0, client1, delivery_location0, supplier_product0, save_objects):
+        client1.save()
+        ordered_product1 = OrderedProduct.order(delivery_location0, client1, supplier_product0, 3)
+        ordered_product2 = OrderedProduct.order(delivery_location0, client0, supplier_product0, 8)
+        assert ordered_product1 in OrderedProduct.objects.all()
+        assert ordered_product2 in OrderedProduct.objects.all()
+        assert ordered_product1.id != ordered_product2.id
```
Comment (Omeramsc) [inline ordered_product/tests.py:82]: This assertion seems redundant. you created those products above, you should be aware of what they consist of.

Comment (Omeramsc) [reply ordered_product/tests.py:82]: (same for all of the assertions below)

File: ordered_product/tests.py [hunk 40]
```diff
@@ -0,0 +1,93 @@
...
+    def test_increase_quantity(self, ordered_product0, save_objects):
+        ordered_product0.quantity = 3
+        ordered_product0.save_ordered_product()
+        ordered_product0.increase_quantity(6)
+        assert ordered_product0.quantity == 9
+        assert ordered_product0 in OrderedProduct.objects.all()
+
+    @pytest.mark.django_db()
+    def test_order(self, client0, client1, delivery_location0, supplier_product0, save_objects):
+        client1.save()
```
Comment (Omeramsc) [inline ordered_product/tests.py:77]: what is the difference between client1 and client zero? why do we need them both?

**Comment by Omeramsc** [comment]
Do note that when the other 2 open PRs will merge, you'll have to rebase and remove the other classes' declarations

File: ordered_product/tests.py (hunk 39 above)
Comment (MeitarRizner) [reply ordered_product/tests.py:82]: Okay. I'll fix it.

File: ordered_product/tests.py (hunk 40 above)
Comment (MeitarRizner) [reply ordered_product/tests.py:77]: I'm understanding it now.. I'll change the test.

File: ordered_product/models.py [hunk 41]
```diff
@@ -0,0 +1,63 @@
...
+            raise e
+        return self
+
+    def delete_ordered_product(self):
+        OrderedProduct.objects.get(id=self.id).delete()
+
+    def decrease_quantity(self, qn):
+        self.quantity -= qn
+        self.save()
+        if (self.quantity <= 0):
```
Comment (yftacherzog) [inline ordered_product/models.py:50]: the `save` line is potentially putting the DB in an invalid state until you delete it. it means a race condition in which others can access invalid data between the time you save it and the time you delete it.

Honestly, any way you turn this around, there's going to be some sort of race condition here unless you make this method atomic (e.g. you decrement a value different than the one you read, etc.), and I don't want you to spend time on that. so, for now, I think you can leave it as-is, but if someone asks you about it one day, just know there's an issue here :)

Comment (yftacherzog) [reply ordered_product/models.py:50]: (All in all, nothing to do here for now, just FYI for whoever reads)

**Review by yftacherzog — APPROVED** [review]
lgtm

**Review by thetaxas2468 — APPROVED** [review]
LGTM!

**Review by Omeramsc — APPROVED** [review]
LGTM, but we'll have to wait for #38 to merge, and rebase this on it, before we'll be able to merge.

**Comment by Omeramsc** [comment]
@MeitarRizner #38 is merged, solve the conflicts and we'll be able to merge it :)

**Comment by MeitarRizner** [comment]
> @MeitarRizner #38 is merged, solve the conflicts and we'll be able to merge it :)

@Omeramsc Done.

**Review by Omeramsc — APPROVED** [review]
Good job! merged
